Util/ai_command.py

```diff
@@ -3,12 +3,58 @@
 from collections import namedtuple
 
 
-class AICommand(namedtuple('AICommand', 'robot_id target kick_type kick_force charge_kick dribbler_active path '
-                                        'cruise_speed end_speed ball_collision')):
 
+class AICommand(namedtuple('AICommand', 'target kick_type kick_force charge_kick dribbler_active '
+                                        'cruise_speed end_speed ball_collision pathfinder_on')):
     __slots__ = ()
 
-    def __new__(cls, robot_id, target=None, kick_type=None, kick_force=0, charge_kick=False, dribbler_active=False,
-                path=None, cruise_speed=1, end_speed=0, ball_collision=True):
-        return super().__new__(cls, robot_id, target, kick_type, kick_force, charge_kick, dribbler_active, path,
-                               cruise_speed, end_speed, ball_collision)
+
+class CmdBuilder:
+
+    def __init__(self):
+        # Does are the defaults values of AICommand
+        # A robot should stay idle with those value
+        self.target = None
+        self.kick_type = None
+        self.kick_force = 0
+        self.charge_kick = False
+        self.dribbler_active = False
+        self.cruise_speed = 0
+        self.end_speed = 0
+        self.ball_collision = True
+        self.pathfinder_on = True
+
+    def withMoveTo(self, target, cruise_speed=1, end_speed=0, ball_collision=True):
+        self.target = target
+        self.cruise_speed = cruise_speed
+        self.end_speed = end_speed
+        self.ball_collision = ball_collision
+        self.pathfinder_on = True
+        return self
+
+    def withKick(self, kick_force=1):
+        self.kick_force = kick_force
+        self.kick_type = 1 # For the moment we only have one type
+        return self
+
+    def withForceDribbler(self):
+        self.dribbler_active = True
+        return self
+
+    def withChargeKicker(self):
+        self.charge_kick = True
+        return self
+
+    def build(self):
+        return AICommand(self.target,
```

Comment:
Can the CmdBuilder be an AICommand which update itself?

Reply:
you can use the nametuple private function `self._replace(...)` inside the class definition.

Reply:
Le but du design pattern est justement d'avoir un état interne qui est construit une seule fois. L'avantage de cela est que l'on peut détecté si deux configurations conflictuelle sont faites. Par exemple, je fais  un withMoveTo et après je fais withMoveAround, ses deux choses sont conflictuelle, il faut donc que la construction génère une erreur.
Mais c'est vrai que CmdBuilder est pas immédiatement relié à une ai_command.
J'aime aussi l'idée que ai_command est immutable.

Reply:
elle serait immuable en dehors de l'intérieur de la classe